msal_extensions/token_cache.py

```diff
@@ -160,15 +156,14 @@ def __init__(self,
             lock_location=lock_location)
         self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
 
-    def _write(self):
+    def _write(self, contents):
         with open(self._cache_location, 'wb') as handle:
```

Comment:
Is there any specific reason that this method does not inherit from its base class? Currently you lost that special exception handling behavior.

Reply:
So, I've been thinking about it, and I think I'd rather let _read and _write throw exceptions if they're not able to complete their function. The caller (`add`, `modify`, and `find` in our case) can decide whether or not it was acceptable to not not be able to perform that function. This change made here: 47e985c899fbfbab9a3f70db3b81a873666c5307

To your broader point though, I think that `FileTokenCache` is the only implementation that writing in plain text UTF-8. I don't think there's much opportunity for code re-use by encouraging folks to call `super` here. Seems like we may just be on different wave-lengths on this. Let me know if there's something I'm missing!

Reply:
Really LOVE this kind of conversation! :-)  I'm going to agree on your latest implementation, meanwhile I'm still providing some thoughts for the sake of mutual learning.

Regarding to putting those exception handling logic out of _read() and _write() and move to their callers, I understand your point and I agree. In 98% of time it would be the correct choice because exception handling should happen more close to the "application" layer, rather than in the "underlying common tool" layer. In this particular case, though, it might fall into the 1% of time when it doesn't really matter, because those two _read() and _write() are not really a common reusable utility to the outside world, they are only needed inside this `FileTokenCache` class family, so it would be fine for them to include some behaviors common to this `FileTokenCache` family, if that means better code reuse. All in all, since the difference here is not that significant and "it doesn't matter", let's settle this case and just use your lastest implementation. :-)

Regarding to that broader point of method inheritance, I know the actual code re-use benefit is not significant in this particular case (but then again, we introduced one more place for that potential "rs" thing :-) ). It is just more of a conceptual convention that a sub-class method is supposed to inherit its base class behavior and build on top of that, whenever possible. Imagine a new maintainer is picking up this method, his first reaction would probably be alerted: "Why does this sub-class method not inherit from the base method? Given that there was no inline comment to confirm this unusual behavior is intended, could this be a bug?", and he would need an investigation to understand what else, if any, is in the base class that could be missing in the subclass; or is it really an intended behavior. All of those were indeed what I did when I was doing this PR review; I would not have to do so if this `WindowsTokenCache._read()` was implemented as a logically self-contained `return self._dp_agent.unprotect(super(..., self)._read())`. (And, for the record, I did same investigation on `OSXTokenCache._read()` and understood why it did not inherit from the base. So that part is fine.)

Resolving this conversation now. :-)